src/mail-app/mail/model/MailListModel.ts

```diff
@@ -0,0 +1,314 @@
+	}
+
+	get stateStream(): Stream<ListState<Mail>> {
+		return this.listModel.stateStream.map((state) => {
+			const items = state.items.map((item) => item.mail)
+			const selectedItems: Set<Mail> = new Set()
+			for (const item of state.selectedItems) {
+				selectedItems.add(item.mail)
+			}
+			const newState: ListState<Mail> = {
+				...state,
+				items,
+				selectedItems,
+			}
+			return newState
+		})
+	}
+
+	isLoadingAll(): boolean {
+		return this.listModel.state.loadingAll
+	}
+
+	isItemSelected(mailId: Id): boolean {
+		const loadedMail = this.mailMap.get(mailId)
+		if (loadedMail == null) {
+			return false
+		}
+		return this.listModel.isItemSelected(getElementId(loadedMail.mailSetEntry))
+	}
+
+	getMail(mailId: Id): Mail | null {
+		return this.getLoadedMail(mailId)?.mail ?? null
+	}
+
+	getMailSetEntry(mailSetEntryId: Id): MailSetEntry | null {
+		const { mailId } = deconstructMailSetEntryId(mailSetEntryId)
+		return this.getLoadedMail(mailId)?.mailSetEntry ?? null
+	}
+
+	loadAndSelect(mailId: Id, shouldStop: () => boolean): Promise<LoadedMail | null> {
+		return this.listModel.loadAndSelect(mailId, shouldStop)
+	}
+
+	onSingleSelection(mail: Mail) {
+		this.listModel.onSingleSelection(assertNotNull(this.getLoadedMail(mail)))
+	}
+
+	selectNone() {
+		this.listModel.selectNone()
+	}
+
+	cancelLoadAll() {
+		this.listModel.cancelLoadAll()
+	}
+
+	async loadInitial() {
+		await this.listModel.loadInitial()
+	}
+
+	getSelectedAsArray(): Array<Mail> {
+		return this.listModel.getSelectedAsArray().map(({ mail }) => mail)
+	}
+
+	async handleEntityUpdate(update: EntityUpdateData) {
+		if (isUpdateForTypeRef(MailFolderTypeRef, update)) {
+			// In case labels change trigger a list redraw.
+			// We need to do it because labels are passed out of band and are not part of the list state.
+			this.reapplyFilter()
+		} else if (isUpdateForTypeRef(MailSetEntryTypeRef, update) && isSameId(this.mailSet.entries, update.instanceListId)) {
+			if (update.operation === OperationType.DELETE) {
+				await this.listModel.deleteLoadedItem(update.instanceId)
+			} else if (update.operation === OperationType.CREATE) {
+				const loadedMail = await this.loadSingleMail([update.instanceListId, update.instanceId])
+				await this.listModel.waitLoad(async () => {
+					if (this.listModel.itemWithinLoadedRange(loadedMail)) {
+						this.listModel.insertLoadedItem(loadedMail)
+					}
+				})
+			}
+		} else if (isUpdateForTypeRef(MailTypeRef, update)) {
+			const mailItem = this.mailMap.get(update.instanceId)
+			if (mailItem != null && update.operation === OperationType.UPDATE) {
+				const newMailData = await this.entityClient.load(MailTypeRef, [update.instanceListId, update.instanceId])
+				// Updating the mail in-place does not require waiting for the underlying list model to finish.
+				// We use Object.assign here to ensure references to the mail now have the new mail data
+				Object.assign(mailItem.mail, newMailData)
```

Comment:
I'm not sure if this optimization is worth it. What if new mail has fewer fields assigned than the old one (probably not possible yet but could). I prefer to not mutate globally-referenced instances without reason (even though that's the case for mails right now).

I also don't understand how it would trigger a list update.

Reply:
yeah, we should check, I thought they would eventually get in sync